### ISSUE TITLE
QEO NDIS Control Path Design

### ISSUE THREAD

**Issue by nibanks** [microsoft/quic-offloads#2]
The NDIS control path (add/remove connections) for QEO will use OIDs to communicate with the NIC. The question we have is what type of OIDs should be used? AFAIU, some OIDs have high latency. If no type of OID is fast enough, perhaps instead of OID to plumb a connection, use a special OOB in first packet.

While adding and removing QEO connections is considered control path, it is directly linked with the data path. You can't send a packet using an offload until you have successfully plumbed the connection. Should the app be expected to wait on the OID for some indeterminate time before sending packets leveraging it? Should it encrypt the packets itself while it waits? Can we just let SW fallback handle the race?

**Comment by MilesPennerIntel** [comment]
I have to agree.  Synchronous OIDs are problematic due to the pending inability.  In imposes a contract that encourages DISPATCH level busy waits.  No matter what, the NIC driver will need to either program the key (depending on implementation) via DMA or MMIO writes to the HW which take time.

**Comment by maolson-msft** [comment]
Thanks Jeff and Miles, that's good feedback. I'll go ahead and plan for it to be Direct.

I'd still love a rough estimate on how long it will take to program the key to the HW though!

**Comment by BorisPis** [comment]
How about the following (fast) control path design. The control path uses the same queues as data packets to post connection state updates to HW; with proper fencing. To avoid races, the OS needs to ensure packets are not posted to other queues until a state update completion is received. This will provide the best performance, and it may even be transferred alongside packet data in a NET_BUFFER_LIST.

Alternatively, it will be easiest to use a regular OID, so that FW can be called while blocking the connection; this will be much slower (milliseconds), but it can make initial development easier.

**Comment by BorisPis** [comment]
> Also, what about local NAT? If a piece of software sits between the local UDP stack and the NIC driver and rewrites the local IP or port, should all of those bits of software be rewritten to also update the QEO control path so that the NAT'd tuples line up with the control path parameters, or would the presence of NATs (or even potential NATs) cause QEO to be disabled?

There are a number of ways to handle NAT with QUIC offloading:
- Add HW steering using the new post-NAT address to point to the QUIC decryption state
- Ignore IP addresses and use only the UDP port and connection ID fields to identify connections -- this is better since it avoids the need to update HW about NAT

**Comment by nibanks** [comment]
For the time being, I'm inclines to say if there is some NAT on the box, let's default to the offload getting disabled. If folks think it's an issue worth discussing right now, I'd ask them to please open up a separate issue.

> Alternatively, it will be easiest to use a regular OID, so that FW can be called while blocking the connection; this will be much slower (milliseconds), but it can make initial development easier.

Since performance is of paramount importance, and since we're likely to end up doing the Windows-side development and testing before HW is ready (unless HW is a lot closer than I think?) we likely won't have an "initial development" phase that overlaps. So I think we'll be going with Direct OIDs.
